File: mRemoteV1/App/App.SupportedCultures.vb
Namespace App
    Public Class SupportedCultures
        Inherits Dictionary(Of String, String)

        Private Sub New()
            Dim CultureInfo As Globalization.CultureInfo
            For Each CultureName As String In My.Settings.SupportedUICultures.Split(",")
                Try
                    CultureInfo = New Globalization.CultureInfo(CultureName.Trim)
                    Add(CultureInfo.Name, CultureInfo.TextInfo.ToTitleCase(CultureInfo.NativeName))
                Catch ex As Exception
                    Debug.Print(String.Format("An exception occurred while adding the culture '{0}' to the list of supported cultures. {1}", CultureName, ex.ToString))
                End Try
            Next
        End Sub

        Private Shared _Instance As SupportedCultures = Nothing
        Public Shared Sub InstantiateSingleton()
            If _Instance Is Nothing Then
                _Instance = New SupportedCultures
            End If
        End Sub

        Public Shared Function IsNameSupported(ByVal CultureName As String) As Boolean
            Return _Instance.ContainsKey(CultureName)
        End Function

        Public Shared Function IsNativeNameSupported(ByVal CultureNativeName As String) As Boolean
            Return _Instance.ContainsValue(CultureNativeName)
        End Function

        Public Shared ReadOnly Property CultureName(ByVal CultureNativeName As String) As String
            Get
                Dim Names(_Instance.Count) As String
                Dim NativeNames(_Instance.Count) As String

                _Instance.Keys.CopyTo(Names, 0)
                _Instance.Values.CopyTo(NativeNames, 0)

                For Index As Integer = 0 To _Instance.Count
                    If NativeNames(Index) = CultureNativeName Then
                        Return Names(Index)
                    End If
                Next

                Throw New System.Collections.Generic.KeyNotFoundException()
            End Get
        End Property

        Public Shared ReadOnly Property CultureNativeName(ByVal CultureName As String) As String
            Get
                Return _Instance.Item(CultureName)
            End Get
        End Property

        Public Shared ReadOnly Property CultureNativeNames() As List(Of String)
            Get
                Dim ValueList As New List(Of String)
                For Each Value As String In _Instance.Values
                    ValueList.Add(Value)
                Next
                Return ValueList
            End Get
        End Property
    End Class
End Namespace

File: mRemoteV1/CS/App/App.SupportedCultures.cs
// VBConversions Note: VB project level imports
using System.Collections.Generic;
using System;
using AxWFICALib;
using System.Drawing;
using System.Diagnostics;
using System.Data;
using AxMSTSCLib;
using Microsoft.VisualBasic;
using System.Collections;
using System.Windows.Forms;
// End of VB project level imports


namespace mRemoteNG.App
{
	public class SupportedCultures : System.Collections.Generic.Dictionary<string, string>
		{
			
			private SupportedCultures()
			{
				System.Globalization.CultureInfo CultureInfo = default(System.Globalization.CultureInfo);
				foreach (string CultureName in My.Settings.Default.SupportedUICultures.Split(","))
				{
					try
					{
						CultureInfo = new System.Globalization.CultureInfo(CultureName.Trim());
						Add(CultureInfo.Name, CultureInfo.TextInfo.ToTitleCase(CultureInfo.NativeName));
					}
					catch (Exception ex)
					{
						Debug.Print(string.Format("An exception occurred while adding the culture \'{0}\' to the list of supported cultures. {1}", CultureName, ex.ToString()));
					}
				}
			}
			
			private static SupportedCultures _Instance = null;
			public static void InstantiateSingleton()
			{
				if (_Instance == null)
				{
					_Instance = new SupportedCultures();
				}
			}
			
			public static bool IsNameSupported(string CultureName)
			{
				return _Instance.ContainsKey(CultureName);
			}
			
			public static bool IsNativeNameSupported(string CultureNativeName)
			{
				return _Instance.ContainsValue(CultureNativeName);
			}
			
			public static string get_CultureName(string CultureNativeName)
			{
				string[] Names = new string[_Instance.Count + 1];
				string[] NativeNames = new string[_Instance.Count + 1];
				
				_Instance.Keys.CopyTo(Names, 0);
				_Instance.Values.CopyTo(NativeNames, 0);
				
				for (int Index = 0; Index <= _Instance.Count; Index++)
				{
					if (NativeNames[Index] == CultureNativeName)
					{
						return Names[Index];
					}
				}
				
				throw (new System.Collections.Generic.KeyNotFoundException());
			}
			
			public static string get_CultureNativeName(string CultureName)
			{
				return _Instance[CultureName];
			}
			
public static List<string> CultureNativeNames
			{
				get
				{
					List<string> ValueList = new List<string>();
					foreach (string Value in _Instance.Values)
					{
						ValueList.Add(Value);
					}
					return ValueList;
				}
			}
		}
}
